Security/Authorization/Expression/ExpressionVoter.php

```diff
@@ -115,17 +115,44 @@ protected function getCompiler()
 
     private function createEvaluator(Expression $expr)
     {
-        if ($this->cacheDir) {
-            if (is_file($file = $this->cacheDir.'/'.sha1($expr->expression).'.php')) {
-                return require $file;
+        if (!$this->cacheDir) {
+            return eval($this->getCompiler()->compileExpression($expr));
+        }
+
+        $file = $this->cacheDir . '/' . $expr->getHashCode() . '.php';
+
+        if (is_file($file)) {
+            $callback = require $file;
+
+            if (is_callable($callback)) {
+                return $callback;
             }
 
-            $source = $this->getCompiler()->compileExpression($expr);
-            file_put_contents($file, "<?php\n".$source);
+            if (null !== $this->logger) {
+                $this->logger->error(sprintf('Expression "%s" failed to load; cache broken', $expr->expression));
+            }
+        }
+
+        if (null !== $this->logger) {
+            $this->logger->info(sprintf('Caching "%s" expression; hashCode: %s', $expr->expression, $expr->getHashCode()));
+        }
+
+        $source = $this->getCompiler()->compileExpression($expr);
+        $content = "<?php\n" . $source;
+
+        // avoid blocking I/O
+        $tmpfile = $this->cacheDir . '/' . uniqid($expr->getHashCode(), true) . '.tmp';
```

Comment:
This should have some random component to it, otherwise we might see the same problem when many tasks are trying to write the same file.